### PR TITLE
chore: bump dependencies

### LINKED ISSUE #134: Clustering warning in sklearn version 1.2
After we upgraded sklearn to version 1.2, there is a warning shown when I run the tests:
```
site-packages\sklearn\cluster\_agglomerative.py:983: FutureWarning: Attribute `affinity` was deprecated in version 1.2 and will be removed in 1.4. Use `metric` instead
```

### LINKED ISSUE #210: Different dependency version matching between conda and python
We may have slightly different dependency matching affecting the package versions that are installed from conda and pip. That means that when following the install practices recommended in our readme and test case after #202, some packages are installed by conda, then uninstalled and reinstalled by pip.

See https://github.com/Reed-CompBio/spras/pull/202#issuecomment-2744386309

### PR DESCRIPTION
**Pull request by tristan-f-r**
Closes #210. Closes #134.

- [x] Depends on #340.

### CODE REVIEW

**Comment by ntalluri** [comment]
Do you have an example of this (the image) that you could add as a comment?

**Comment by tristan-f-r** [comment]
NixOS 25.05.20250108.bffc22e (Warbler) on x86_64, though I do think this issue is reproducing on CI, so I doubt this is platform dependent. I'll reproduce it once I switch from working on https://github.com/Murali-group/halp/pull/34 for #155.

**Comment by tristan-f-r** [comment]
@ntalluri (for https://github.com/tristan-f-r/spras/blob/54bd449d376756cb5d49785c9cf85e2d7c17aa21/test/ml/test_ml.py#L83-L95):

```
Got:
      algorithm      PC1      PC2
0  test-data-s1 -2.00665 -0.98659
1  test-data-s2 -1.52765  1.07995
2  test-data-s3  3.53430 -0.09336       algorithm      PC1      PC2
0  test-data-s1 -2.00665 -0.98659
1  test-data-s2 -1.52765  1.07995
2  test-data-s3  3.53430 -0.09336

Expected:
      algorithm      PC1      PC2
0  test-data-s1  2.00665 -0.98659
1  test-data-s2  1.52765  1.07995
2  test-data-s3 -3.53430 -0.09336       algorithm      PC1      PC2
0  test-data-s1 -2.00665 -0.98659
1  test-data-s2 -1.52765  1.07995
2  test-data-s3  3.53430 -0.09336
```

**Comment by ntalluri** [comment]
Based on what I’m reading, bumping the version is a good idea because it improves PCA in a meaningful way. Before, PCA signs were chosen by looking at how the data looked after projection (transformed data), but now they are chosen by inspecting each component vector directly. Previously, each solver decided the signs in its own way, which meant that changing the solver could cause components to unexpectedly flip signs. While the mathematical results seem to always still be correct, this causes confusion.  The new approach in version 1.7 makes the sign choice independent of the solver because it relies only on the component vector’s own values, ensuring we always get consistent component signs no matter which solver we use.

We don't allow a user to pick a solver, we have it set to auto at the moment.
https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA:~:text=svd_solver%7B%E2%80%98auto%E2%80%99%2C%20%E2%80%98full,optionally%20truncated%20afterwards.

**Comment by tristan-f-r** [comment]
[Let me know if this commit is the right way to approach that.](https://github.com/Reed-CompBio/spras/pull/310/commits/ae02c993b8fe7fff590d3cf17784a46f4135dd13) To my understanding, this does mess with our sampling test because our component vectors differ per run; hopefully, the aforementioned commit addresses that.

File: test/ml/test_ml.py [hunk 1]
```diff
@@ -102,7 +106,7 @@ def test_pca_robustness(self):
             coord = coord.round(5)  # round values to 5 digits to account for numeric differences across machines
             coord.sort_values(by='algorithm', ignore_index=True, inplace=True)
 
-            assert coord.equals(expected)
+            assert coord.equals(expected) or coord.equals(expected_other)
```
Comment (ntalluri) [inline test/ml/test_ml.py:109]: Is there 2 possible outputs?

**Comment by ntalluri** [comment]
> [Let me know if this commit is the right way to approach that.](https://github.com/Reed-CompBio/spras/pull/310/commits/ae02c993b8fe7fff590d3cf17784a46f4135dd13) To my understanding, this does mess with our sampling test because our component vectors differ per run; hopefully, the aforementioned commit addresses that.

For that specific test, we want the output to be the same despite shuffling the rows and columns. So having 2 different outputs is not what we would want.

File: test/ml/test_ml.py (hunk 1 above)
Comment (agitter) [reply test/ml/test_ml.py:109]: I believe the answer is yes? It should be valid to flip the solution.

**Comment by tristan-f-r** [comment]
> Can you help me get caught up on the core question? It seems valid to have a transformation of the PCA solution where the signs are flipped.

The essence of the test that was broken here was row shuffling preserving PCA output:

```py
dataframe_shuffled = dataframe.sample(frac=1, axis=1)  # permute the columns
ml.pca(dataframe_shuffled, OUT_DIR + 'pca-shuffled-columns.png', OUT_DIR + 'pca-shuffled-columns-variance.txt',
    OUT_DIR + 'pca-shuffled-columns-coordinates.tsv')
coord = pd.read_table(OUT_DIR + 'pca-shuffled-columns-coordinates.tsv')
coord = coord.round(5)  # round values to 5 digits to account for numeric differences across machines
coord.sort_values(by='algorithm', ignore_index=True, inplace=True)
```

This now needs an assertion check for two possible signings:

```
assert coord.equals(expected) or coord.equals(expected_other)
```

File: test/ml/expected/expected-pca-coordinates-2.tsv [hunk 2]
```diff
@@ -0,0 +1,4 @@
+algorithm	PC1	PC2
```
Comment (agitter) [inline test/ml/expected/expected-pca-coordinates-2.tsv:1]: If these two files are the same except for the PC1 sign, do we want them to be stored separately? Or is it better to track a single file and have the test make a copy and negate PC1?

Comment (agitter) [reply test/ml/expected/expected-pca-coordinates-2.tsv:1]: Fair point. I like `negated` better.

**Review by agitter — APPROVED** [review]
I haven't run locally in the new environment, but all the changes look okay.

**Comment by ntalluri** [comment]
If it is the ensembling one, there was a data.pickle I added that has a toy dataset network that is needed for ensembling in general that we use to ensure we calculate recall correct. I'm not sure why that would need to be updated (if it is this pickled dataset) because it has nothing to do with the pca.

**Comment by agitter** [comment]
> Regardless, we probably should stop committing artifacts to main anyway - this causes problems; see https://github.com/Reed-CompBio/spras/pull/320 for an example of a large test refactor because of a related issue

This seems worthwhile to discuss in a new issue as a testing strategy change if it affects multiple tests.

**Comment by tristan-f-r** [comment]
https://github.com/Reed-CompBio/spras/issues/339

**Comment by tristan-f-r** [comment]
> @tristan-f-r Which PR was it? The ensembling one?

It was #212 ([origin commit](https://github.com/Reed-CompBio/spras/commit/c3fe362d16371421a78d25e4480d92b6c6c77f9b)).

**Comment by tristan-f-r** [comment]
> If it is the ensembling one, there was a data.pickle I added that has a toy dataset network that is needed for ensembling in general that we use to ensure we calculate recall correct. I'm not sure why that would need to be updated (if it is this pickled dataset) because it has nothing to do with the pca.

Pickled objects contain internal representations of the object, and are subject to change throughout package versions. In this case, `pandas` itself changed.

**Comment by tristan-f-r** [comment]
https://github.com/Reed-CompBio/spras/pull/340

**Review by ntalluri — DISMISSED** [review]

File: test/ml/expected/expected-pca-coordinates-negated.tsv
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply test/ml/expected/expected-pca-coordinates-negated.tsv:1]: It only negated one column, like before.

File: test/ml/test_ml.py [hunk 3]
```diff
@@ -88,6 +88,7 @@ def test_pca_remove_empty_pathways(self):
         coord = pd.read_table(OUT_DIR + 'pca-coordinates.tsv')
         coord = coord.round(5)  # round values to 5 digits to account for numeric differences across machines
         expected = pd.read_table(EXPECT_DIR + 'expected-pca-coordinates.tsv')
+        pd.read_table(EXPECT_DIR + 'expected-pca-coordinates-negated.tsv')
```
[the remark this answers is not in the document]
Comment (tristan-f-r) [reply test/ml/test_ml.py:91]: It isn't 👍

**Comment by ntalluri** [comment]
The code itself looks good to me. I still need to test the new environment updates locally.

**Review by ntalluri — COMMENTED** [review]
Code looks good to me. However, I am going to test the environment updates locally before approving.

**Comment by tristan-f-r** [comment]
Okay - sorry about this terribly long hill of a PR for what it was supposed to be. That should be the last commit which fixes the tests from the latest merge :+1:

**Review by agitter — APPROVED** [review]
I ran the `TestML` tests locally and they passed
